### ISSUE TITLE
Middleware validates multiples times with same validator

### ISSUE THREAD

**Issue by ezypeeze** [benmccallum/fairybread#36]
Hey there,

I got these weird issue when having multiple inputs, each one has only one validator.
![image](https://user-images.githubusercontent.com/10672637/107097945-f9ccb600-6805-11eb-8b99-54c1fd12e7cd.png)

I did some analysis on the code and found a little bug on the `DefaultValidatorProvider`.

I got a quick fix it for it:
```
if (!Cache.TryGetValue(validatedType, out var validatorsForType))
{
    Cache[validatedType] = validatorsForType = new List<ValidatorDescriptor>();
}

// here we need to check if the validator type is already included in the validatorForType (the cache dictionary)
if (!validatorsForType.Any(x => x.ValidatorType.Equals(validatorType)))
{
    var requiresOwnScope = ShouldBeResolvedInOwnScope(validatorType);
    var validatorDescriptor = new ValidatorDescriptor(validatorType, requiresOwnScope);

    validatorsForType.Add(validatorDescriptor);
}
```

I would make a PR myself, but going to be off the whole weekend.

**Comment by benmccallum** [comment]
Wow, that does look like a fun bug :P I'll try get this sort in the next day or so, worst case it'll be this weekend. Thanks for reporting it and the investigation! 🙌

**Comment by benmccallum** [comment]
Hi @ezypeeze , I'm having problems repro'ing this one. 
* I thought it was just using multiple root-level fields at once, nope.
* Then that it might be multiple queries fired over time, nope (since that validator discovery only happens once at startup).
* Also thought that the cache could be a static var and if you registered the `DefaultValidatorProvider` as non-singleton it could multi-register but it's not static so that should be OK.

The only other things I can think ofare:
* could you be multi-registering your FluentValidator validators in your Startup? 
* your validators overlapping in some way?

Do you think you could let me know what you're setup is or even better do a small repro in a test? I started a test in `InputValidationMiddleware::Multi_TopLevelFields_And_MultiRuns_Works` where I'm trying to repro which you could use as a starting point.

Cheers!

**Comment by ezypeeze** [comment]
Hey @benmccallum, here it is (using v4 - nice work removing the duplicated work for registry the validators 👍)

[fairy-bread-example.zip](https://github.com/benmccallum/fairybread/files/5988449/fairy-bread-example.zip)

![image](https://user-images.githubusercontent.com/10672637/108061854-d3521a80-7050-11eb-908b-2e999877751d.png)

---

Instead of giving just an element in the errors array, it gives the number of registered validators I've (in this case I've just two - CreateBookInput.Validator & CreateAuthorInput.Validator)

**Comment by benmccallum** [comment]
Hi @ezypeeze, 

Thanks for this and sorry it's taken me a while to have a look at it. 

So I've had a look and the problem is that you're registering your validators twice.
```csharp
        public void ConfigureServices(IServiceCollection services)
        {
            services
                .AddValidatorsFromAssemblyContaining<CreateAuthorInput.Validator>()
                .AddValidatorsFromAssemblyContaining<CreateBookInput.Validator>();
```

Crucially, the method you're using from FluentValidation.DependencyInjectionExtensions to register is `AddValidatorsFromAssemblyContaining`, which means it'll take the type you've passed, determine it's assembly, then scan that assembly for all validators and register each one of them. So if you call it like above for each validator you have, which are both in the same assembly, you'll double-register all validators in that assembly. 

That would also explain why you were seeing like 5 duplicate errors in your original post, cuz you were probably calling it for say 5 validators you were trying to register in the same assembly. And also why I couldn't reproduce the issue :P 

So basically just call that method once for each assembly you wanna register types inside and you should be good.

I'm not sure if there's anything I can really do to protect against this in FairyBread... the same behavior would happen if you were using FluentValidation in any other context, like MVC or if you called Validate manually. There may be a legitimate reason for this behavior, not that I can think of any, so although I could de-dupe in FairyBread I'm hesitant to do so.

Will close this but feel free to re-open if I've made a mistake.

**Comment by ezypeeze** [comment]
Hey @benmccallum,

```csharp
collection.AddValidatorsFromAssemblyContaining<Startup>();
```

did the trick! (since I have all types in same assembly).

Thanks for the help!
